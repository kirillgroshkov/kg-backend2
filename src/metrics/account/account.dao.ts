import { BaseDatastoreDao } from '@src/db/datastore/base.datastore.dao'
import {
  AccountBM,
  accountBMSchema,
  AccountDBM,
  accountDBMSchema,
  AccountFM,
  accountFMSchema,
} from '@src/metrics/account/account.model'
import { securityService } from '@src/srv/security.service'

export class AccountDao extends BaseDatastoreDao<AccountBM, AccountDBM, AccountFM> {
  KIND = 'Account'
  excludeFromIndexes = []
  BACKEND_RESPONSE_PROPERTY = 'account'
  DBM_SCHEMA = accountDBMSchema
  BM_SCHEMA = accountBMSchema
  FM_SCHEMA = accountFMSchema

  async getByAccountKey (key?: string): Promise<AccountBM | undefined> {
    if (!key) return
    const keyHash = securityService.hashPassword(key)
    return this.findOneBy('keyHash', keyHash)
  }
}
