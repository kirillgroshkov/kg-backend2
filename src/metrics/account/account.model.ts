import {
  idSchema,
  objectSchema,
  stringSchema,
  unixTimestampSchema,
} from '@naturalcycles/nodejs-lib'
import { BaseDBEntity } from '@src/db/datastore/datastore.model'

export interface AccountBM extends BaseDBEntity {
  name: string
  keyHash: string
}

export interface AccountDBM extends AccountBM {}

export const accountBMSchema = objectSchema({
  id: idSchema,
  created: unixTimestampSchema,
  updated: unixTimestampSchema,
  name: stringSchema,
  keyHash: stringSchema,
})

export const accountDBMSchema = accountBMSchema
