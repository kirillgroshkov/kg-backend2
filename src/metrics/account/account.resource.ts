import { ReqRouter } from '@src/server/req.router'
import { accountDao } from '@src/services'

const router = new ReqRouter()
export const accountResource = router.resource

router.get('/', async (req, res) => {
  const accounts = await accountDao.getAll()

  res.json(accounts)
})
