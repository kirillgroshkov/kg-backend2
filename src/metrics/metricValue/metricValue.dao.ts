import { BaseDatastoreDao } from '@src/db/datastore/base.datastore.dao'
import {
  MetricValueBM,
  metricValueBMSchema,
  MetricValueDBM,
  metricValueDBMSchema,
} from '@src/metrics/metricValue/metricValue.model'

export class MetricValueDao extends BaseDatastoreDao<MetricValueBM, MetricValueDBM> {
  KIND = 'MetricValue'
  excludeFromIndexes = []
  BACKEND_RESPONSE_PROPERTY = 'metricValue'
  DBM_SCHEMA = metricValueDBMSchema
  BM_SCHEMA = metricValueBMSchema

  // convenience method
  async get (accountId: string, metricId: string, ts: number): Promise<MetricValueBM | undefined> {
    return this.getById(this.naturalId(accountId, metricId, ts))
  }

  async beforeBMToDBM (bm: MetricValueBM): Promise<MetricValueDBM> {
    return {
      ...bm,
      id: this.naturalId(bm.accountId, bm.metricId, bm.ts),
    }
  }

  async beforeDBMToBM (dbm: MetricValueDBM): Promise<MetricValueBM> {
    return {
      ...dbm,
      ...this.parseNaturalId(dbm.id),
    }
  }

  naturalId (accountId: string, metricId: string, ts: number): string {
    return [accountId, metricId, ts].join('_')
  }

  parseNaturalId (id: string): { accountId: string; metricId: string; ts: number } {
    const [accountId, metricId, ts] = id.split('_')

    return {
      accountId,
      metricId,
      ts: Number(ts),
    }
  }
}
