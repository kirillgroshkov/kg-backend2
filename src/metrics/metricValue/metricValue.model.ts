import {
  anyObjectSchema,
  idSchema,
  numberSchema,
  objectSchema,
  stringSchema,
  unixTimestampSchema,
} from '@naturalcycles/nodejs-lib'
import { CreatedUpdatedVer } from '@src/db/datastore/datastore.model'

export interface MetricValueBM extends CreatedUpdatedVer {
  accountId: string
  metricId: string
  ts: number
  v: number
  meta?: any
  url?: string
}

export interface MetricValueDBM extends MetricValueBM {
  /**
   * ${accountId}_${metricId}_${ts}
   */
  id: string
}

export const metricValueBMSchema = objectSchema({
  created: unixTimestampSchema,
  updated: unixTimestampSchema,
  accountId: idSchema,
  metricId: stringSchema, // slugSchema
  ts: unixTimestampSchema,
  v: numberSchema,
  meta: anyObjectSchema.optional(),
  url: stringSchema.optional(),
})

export const metricValueDBMSchema = metricValueBMSchema.concat(
  objectSchema({
    id: stringSchema,
  }),
)

export interface MetricValueInput {
  metricId: string
  ts: number
  v: number
  meta?: any
  url?: string
}

export const metricValueInputSchema = objectSchema({
  metricId: idSchema,
  ts: unixTimestampSchema,
  v: numberSchema,
  meta: anyObjectSchema.optional(),
  url: stringSchema.optional(),
})
