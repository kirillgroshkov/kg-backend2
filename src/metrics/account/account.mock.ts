import { AccountBM } from '@src/metrics/account/account.model'
import { X_ACCOUNT_KEY } from '@src/server/requestContext.service'
import { securityService } from '@src/srv/security.service'
import { MOCK_TS } from '@src/test/mock/mock.cnst'

export const MOCK_ACCOUNT1_KEY = '123456'
export const MOCK_ACCOUNT1_KEY_HASH = securityService.hashPassword(MOCK_ACCOUNT1_KEY)
export const MOCK_ACCOUNT1_HEADERS = {
  [X_ACCOUNT_KEY]: MOCK_ACCOUNT1_KEY,
}

export function mockAccount1 (): AccountBM {
  return {
    id: 'mockorg1',
    created: MOCK_TS,
    updated: MOCK_TS,
    name: 'Org1',
    keyHash: MOCK_ACCOUNT1_KEY_HASH,
  }
}
