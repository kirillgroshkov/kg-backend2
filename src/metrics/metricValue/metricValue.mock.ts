import { MetricValueBM, MetricValueInput } from '@src/metrics/metricValue/metricValue.model'
import { MOCK_TS } from '@src/test/mock/mock.cnst'
import { mockAccount1 } from '../account/account.mock'

export function mockMetricValue1 (): MetricValueBM {
  const acc = mockAccount1()

  return {
    accountId: acc.id,
    metricId: 'metric1',
    created: MOCK_TS,
    updated: MOCK_TS,
    ts: MOCK_TS,
    v: 10,
  }
}

export function mockMetricValueInput1 (): MetricValueInput {
  return {
    metricId: 'metric1',
    ts: MOCK_TS,
    v: 10,
  }
}

export function mockMetricValueInput1_2 (): MetricValueInput {
  return {
    metricId: 'metric2',
    ts: MOCK_TS + 2,
    v: 11,
  }
}
