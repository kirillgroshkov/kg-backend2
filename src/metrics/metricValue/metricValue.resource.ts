import { metricValuePutHandler } from '@src/metrics/metricValue/handlers/metricValue.put.handler'
import { metricValueInputSchema } from '@src/metrics/metricValue/metricValue.model'
import { reqValidationMiddleware } from '@src/server/handlers/reqValidation.mw'
import { ReqRouter } from '@src/server/req.router'

const router = new ReqRouter()
export const metricValueResource = router.resource

router.put('/', reqValidationMiddleware('body', metricValueInputSchema), metricValuePutHandler)
