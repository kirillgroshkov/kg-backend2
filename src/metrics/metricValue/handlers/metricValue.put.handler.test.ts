import { MetricValueInput } from '@src/metrics/metricValue/metricValue.model'
import { metricValueResource } from '@src/metrics/metricValue/metricValue.resource'
import { metricValueDao } from '@src/services'
import { MOCK_ACCOUNT1_HEADERS, mockAccount1 } from '@src/test/mock/account.mock'
import { mockDBState1 } from '@src/test/mock/db.mock'
import { mockMetricValueInput1 } from '@src/test/mock/metricValue.mock'
import { resourceTestService } from '@src/test/resource.test.service'
import { CREATED_UPDATED_MATCHERS } from '@src/test/test.util'

beforeAll(async () => {
  await mockDBState1()
})

const app = resourceTestService.createAppWithResource(metricValueResource)
const request = (input: MetricValueInput) =>
  app
    .put('/')
    .send(input)
    .set(MOCK_ACCOUNT1_HEADERS)

test('PUT /metricvalues', async () => {
  const input = mockMetricValueInput1()
  const acc = mockAccount1()

  const { body, status } = await request(input)
  console.log(status)
  console.log(body)

  expect(status).toBe(201)
  expect(body).toEqual({})

  // Should save to DB
  const id = metricValueDao.naturalId(acc.id, input.metricId, input.ts)
  expect(await metricValueDao.getById(id)).toMatchSnapshot(CREATED_UPDATED_MATCHERS)
})
