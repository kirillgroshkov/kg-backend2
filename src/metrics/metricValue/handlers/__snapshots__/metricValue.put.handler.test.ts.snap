// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PUT /metricvalues 1`] = `
Object {
  "accountId": "mockorg1",
  "created": Any<Number>,
  "metricId": "metric1",
  "ts": 1529539200,
  "updated": Any<Number>,
  "v": 10,
}
`;
